### DISCUSSION TITLE
Toolpad takes a new direction

### DISCUSSION THREAD

**Discussion by Janpot** [mui/toolpad#1748] [Announcements]
For the past year that we've been building Toolpad, we've also kept talking to potential users. Having understood their needs better, now it's time to align Toolpad with the insights we've gained over the past months.

## What will change

Toolpad, as a dev tool, will no longer be a hosted solution. You will install the Toolpad editor as a regular npm dependency and run it locally from within your project folder.

Toolpad applications are stored in configuration files within your project folder, so Toolpad will no longer require access to a database to store them. You will be able to hand-edit these configuration files if necessary and the change will immediately be reflected in the visual editor.

Toolpad integrates more closely with Node.js backends. You can expose your data to Toolpad by writing regular JavaScript functions, and these functions will be able to use any npm package, and have access to environment variables available in your project. As a result, Toolpad is now a more code-friendly way of building applications.

## Why are we changing it?

In contrast to traditional low-code tools (e.g. Pipedream), we want to provide more freedom in the areas we don't consider our core functionality. This means you'll be able to use your code editor, version control system, and deployment infrastructure of choice, all while gaining access to simpler workflows to build UIs through a visual editor. The editor allow you to lay out the UI and configure data fetching.

We disagree with the lock-in and poor DX of other low-code tool, therefore we want to provide powerful built-in customization options. We aim to enable a granular transition to pro-code (e.g. Material UI) whenever the low-code approach falls short.

## How do I get started?

Run
```
npx create-toolpad-app
```
And follow the instructions. More info in our [docs](https://mui.com/toolpad/getting-started/quickstart/)

**Comment by panzacoder** [comment]
Interested to see these changes come into being. Any sort of roadmap for the project created yet or still in early planning stages?

**Reply by Janpot** [reply]
The current state of the repo is a functional prototype. We're using it ourselves in production in a few select places (e.g. a bundle size [dashboard](https://tools-public.onrender.com/prod/pages/h71gdad?circleCIBuildNumber=500402&baseRef=master&baseCommit=72af077b116aa5c3dd4b2781023ec0fea95df7e0&prNumber=36311) for the MUI core repository). Our current focus is around:
- Extensibility on the frontend. i.e. bring your own components
- Connecting data from the backend. i.e. plug in to your existing code base
 Some of the features are incomplete or not 100% functional. If you have specific use-cases or would like weight in on the roadmap, we'd love to hear your suggestions.

**Reply by panzacoder** [reply]
Thanks for this! I would have to dig further into how you plan on implementing the backend connections, but with the two features you suggested, I could also see (with some grunt work on my part) this being a visual layer between a headless CMS and components I create. Could become pretty powerful in that respect. I know there are other tools more directly targeted for that use-case, but I really enjoy the DX of being in the MUI ecosystem of things.

**Comment by erikrenaud** [comment]
I`d love top help bringing this product to the dotnetcore plateform with an elegant integration... I have usecases for this for many clients !
